system_tests/batch_size_limit_test.go

```diff
@@ -0,0 +1,145 @@
+	ctx, cancel := context.WithCancel(context.Background())
+
+	builder := NewNodeBuilder(ctx).DefaultConfig(t, true)
+	builder.nodeConfig.BatchPoster.Enable = false
+	builder.L2Info.GenerateGenesisAccount(SenderAccount, big.NewInt(1e18))
+	builder.L2Info.GenerateGenesisAccount(ReceiverAccount, big.NewInt(0))
+
+	if setHighLimit {
+		builder.chainConfig.ArbitrumChainParams.MaxUncompressedBatchSize = NewUncompressedSizeLimit
+	}
+
+	cleanup := builder.Build(t)
+
+	return builder, ctx, func() {
+		cancel()
+		cleanup()
+	}
+}
+
+// buildBigBatch builds a batch that:
+// - consists of a valid transfer tx followed by highly compressible trash data
+// - has an uncompressed size larger than DefaultMaxUncompressedBatchSize but less than NewUncompressedSizeLimit
+// - has a compressed size smaller than the allowed calldata batch size for the test batch poster
+// - is already compressed and has the appropriate header byte
+func buildBigBatch(t *testing.T, l2Info *BlockchainTestInfo) []byte {
+	batchBuffer := bytes.NewBuffer([]byte{})
+
+	// 1. The first tx in the batch is a standard transfer tx used as an indicator whether the batch was processed successfully.
+	standardTx := l2Info.PrepareTx(SenderAccount, ReceiverAccount, 1000000, big.NewInt(TransferAmount), []byte{})
+	err := writeTxToBatch(batchBuffer, standardTx)
+	Require(t, err)
+
+	// 2. The rest of the batch is filled with highly compressible trash data.
+	batchBuffer.Write(bytes.Repeat([]byte{0xff}, params.DefaultMaxUncompressedBatchSize))
+
+	// 3. Compress the batch (as the batch poster would do).
+	compressed, err := arbcompress.CompressWell(batchBuffer.Bytes())
+	Require(t, err)
+
+	// 4. Ensure compressed and uncompressed sizes are as expected.
+	uncompressedSize, compressedSize := len(batchBuffer.Bytes()), len(compressed)
+	require.Greater(t, uncompressedSize, params.DefaultMaxUncompressedBatchSize)
+	require.Less(t, uncompressedSize, NewUncompressedSizeLimit)
+	require.Less(t, compressedSize, arbnode.TestBatchPosterConfig.MaxCalldataBatchSize)
+
+	// 5. Return the compressed batch with the appropriate header byte.
+	return append([]byte{daprovider.BrotliMessageHeaderByte}, compressed...)
+}
+
+// postBatch posts the given batch directly to the L1 SequencerInbox contract.
+func postBatch(t *testing.T, ctx context.Context, builder *NodeBuilder, batch []byte) {
+	seqNum := new(big.Int).Lsh(common.Big1, 256)
+	seqNum.Sub(seqNum, common.Big1)
+
+	seqInboxAddr := builder.L1Info.GetAddress("SequencerInbox")
+	seqInbox, err := bridgegen.NewSequencerInbox(seqInboxAddr, builder.L1.Client)
+	Require(t, err)
+
+	sequencer := builder.L1Info.GetDefaultTransactOpts("Sequencer", ctx)
+
+	tx, err := seqInbox.AddSequencerL2BatchFromOrigin8f111f3c(&sequencer, seqNum, batch, big.NewInt(1), common.Address{}, big.NewInt(0), big.NewInt(0))
+	Require(t, err)
+	_, err = EnsureTxSucceeded(ctx, builder.L1.Client, tx)
```

Comment:
here, you can look at the receipt, search for the event SequencerBatchDelivered and extract from it seqMessageIndex. Then you can wait for this specific batch to be received instead of waiting for a new batch to be received